modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationMessageProcessor.java

```diff
@@ -120,41 +125,54 @@ public OperationMessageProcessor(ExtensionModel extensionModel,
+        .build();
+    classLoader = getExtensionClassLoader();
   }
 
   @Override
   public Event process(Event event) throws MuleException {
-    return withContextClassLoader(getExtensionClassLoader(), () -> {
-      Optional<ConfigurationInstance> configuration = getConfiguration(event);
+    try {
+      return just(event).transform(this).subscribe().block();
+    } catch (Exception e) {
+      throw rxExceptionToMuleException(e);
+    }
+  }
 
-      ComponentIdentifier operationIdentifier =
-          new ComponentIdentifier.Builder().withName(operationModel.getName())
-              .withNamespace(extensionModel.getName().toLowerCase()).build();
+  @Override
+  public Publisher<Event> apply(Publisher<Event> publisher) {
+    return from(publisher).concatMap(checkedFunction(event -> withContextClassLoader(classLoader, () -> {
+      Optional<ConfigurationInstance> configuration = getConfiguration(event);
 
       Map<String, Object> operationParameters = resolverSet.resolve(event).asMap();
-
-      OperationExecutionFunction operationExecutionFunction = (parameters, operationEvent) -> {
-        ExecutionContextAdapter operationContext = createExecutionContext(configuration, parameters, operationEvent);
-        MuleEvent muleEvent = doProcess(event, operationContext);
-        return (Event) muleEvent;
-      };
-
-      OperationPolicy policy =
-          policyManager.createOperationPolicy(operationIdentifier, event, operationParameters, operationExecutionFunction);
-      return policy.process(event);
+      ExecutionContextAdapter operationContext = createExecutionContext(configuration, operationParameters, event);
+
+      if (operationModel.isBlocking()) {
+        OperationExecutionFunction operationExecutionFunction =
+            (parameters, operationEvent) -> doProcess(operationEvent, operationContext).block();
+        OperationPolicy policy =
+            policyManager.createOperationPolicy(operationIdentifier, event,
+                                                operationParameters,
+                                                operationExecutionFunction);
+        return justOrEmpty(policy.process(event));
+      } else {
+        return doProcess(event, operationContext);
+      }
     }, MuleException.class, e -> {
       throw new DefaultMuleException(e);
-    });
+    })));
   }
 
-  protected org.mule.runtime.api.message.MuleEvent doProcess(Event event, ExecutionContextAdapter operationContext)
+  protected Mono<Event> doProcess(Event event, ExecutionContextAdapter operationContext)
       throws MuleException {
-    Object result = executeOperation(operationContext, event);
-    MuleEvent resultEvent = returnDelegate.asReturnValue(result, operationContext);
-    return resultEvent;
+    return executeOperation(operationContext, event)
+        .map(value -> returnDelegate.asReturnValue(value, operationContext))
+        .otherwiseIfEmpty(justOrEmpty(returnDelegate.asReturnValue(null, operationContext)));
```

Comment:
Don't you need to return original event if there was no result from the operation rather than `returnDelegate.asReturnValue(null, operationContext))`?